### ISSUE TITLE
[MSHADE-97] Add the ability to specify "entry points" for minimizeJar 

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-shade-plugin#477]
**[Jared Bunting](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=deraj123)** opened **[MSHADE-97](https://issues.apache.org/jira/browse/MSHADE-97?redirect=false)** and commented

Currently, the "minimizeJar" option assumes that you want to include all classes in the current project and those that they depend on.  Sometimes, classes are accessed via reflection, or some other means, and therefore don't get included.  The ability to specify additional entry points (as classes) to this dependency tree would be useful.

I have added this as a patch that I am attaching.

---

**Affects:** 1.4

**Attachments:**
- [entrypoints.patch](https://issues.apache.org/jira/secure/attachment/12718939/entrypoints.patch) (_2.87 kB_)
- [entrypoints.patch](https://issues.apache.org/jira/secure/attachment/12718930/entrypoints.patch) (_2.72 kB_)
- [entrypoints-2.1.patch](https://issues.apache.org/jira/secure/attachment/12719006/entrypoints-2.1.patch) (_3.85 kB_)
- [entrypoints-2.1.patch](https://issues.apache.org/jira/secure/attachment/12719001/entrypoints-2.1.patch) (_2.69 kB_)

4 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Jared Bunting](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=deraj123)** commented

Correcting formatting in patch.

**Comment by jira-importer** [comment]
**[Benson Margulies](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bmargulies)** commented

Could I interest you in adding a test?

**Comment by jira-importer** [comment]
**[Jared Bunting](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=deraj123)** commented

I'd be happy to, but it may take a day or two.

**Comment by jira-importer** [comment]
**[Stephen Buergler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sixcorners)** commented

Unrelated but.. How many jar minimizers does there need to be in the world? http://proguard.sourceforge.net/#alternatives.html

**Comment by jira-importer** [comment]
**[Dmitry Katsubo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dma_k)** commented

I would like to vote for this feature. There are many minimizers, but not many are present as maven plugin, and not many are easy to configure. Few lines of code which bring a big benefit ? I think it's worth.

**Comment by jira-importer** [comment]
**[Dmitry Katsubo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dma_k)** commented

If somebody is interested, I attach a patch against v2.1.

**Comment by jira-importer** [comment]
**[Dmitry Katsubo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dma_k)** commented

Another version, this time with mask support:

```
<entryPoints>
	<!-- Classes from javax:jai-imageio only to read image formats -->
	<entryPoint>com.sun.media.imageioimpl.plugins.*ImageReader*</entryPoint>
</entryPoints>
```

**Comment by jira-importer** [comment]
**[Juan Lopes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=juanplopes)** commented

I really would like to see this released.

I implemented something like this, but using the filters, adding all transitive dependencies as well. I had some problems computing the transitive dependencies before all artifacts were loaded into the test classpath, some classes just got removed because they didn't appear in the transitive dependencies.
